### PR TITLE
Node system :Controller service

### PR DESCRIPTION
**Pull request by Brackleycassinga**
This pull request adds a configurator system and controller service to Ukama systems to handle  rebooting actions for both the node and site, as well as a REST endpoint on the API gateway level.

### CODE REVIEW

File: systems/node/api-gateway/pkg/rest/router_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+	// act
+	r.ServeHTTP(w, req)
+
+	// assert
+	assert.Equal(t, http.StatusOK, w.Code)
+	c.AssertExpectations(t)
+}
+
+// func TestRouter_PingRoute(t *testing.T) {
```
Comment (vthakur7f) [inline systems/node/api-gateway/pkg/rest/router_test.go:84]: Remove unwanted  code. Missing test case for restart site

File: systems/node/api-gateway/Int.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,4 @@
+FROM 003664043471.dkr.ecr.us-east-1.amazonaws.com/e2e-base-image:latest
+
+ENV PROJECT_NAME=api-gateway
+COPY bin/integration /usr/bin/integration
```
Comment (vthakur7f) [inline systems/node/api-gateway/Int.Dockerfile:4]: Missing integartion test.

File: systems/node/controller/pb/controller.proto [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+// Defines the service for controller operations
+service ControllerService {
+  rpc RestartSite(RestartSiteRequest) returns (RestartSiteResponse);
+  rpc RestartNode(RestartNodeRequest) returns (RestartNodeResponse);
+}
+
+
+message RestartSiteResponse {
+  RestartStatus status = 1;
+
```
Comment (vthakur7f) [inline systems/node/controller/pb/controller.proto:17]: empty line

File: systems/node/controller/pb/controller.proto [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+message RestartNodeRequest {
+    string nodeId = 1 [(validator.field) = {uuid_ver: 4, string_not_empty: true}, json_name = "node_id"];;
+   }
+
+message RestartSiteRequest {
+    string siteName =1;
+    string networkId = 2 [(validator.field) = {uuid_ver: 4, string_not_empty: true}, json_name = "network_id"];
+
```
Comment (vthakur7f) [inline systems/node/controller/pb/controller.proto:27]: empty line

File: systems/node/controller/pb/controller.proto [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+message RestartNodeResponse {
+    RestartStatus status = 1;
+
```
Comment (vthakur7f) [inline systems/node/controller/pb/controller.proto:32]: empty line

File: systems/node/controller/pb/controller.proto [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+   }
+
+enum RestartStatus {
+    RESTART_STATUS_UNKNOWN = 0;
```
Comment (vthakur7f) [inline systems/node/controller/pb/controller.proto:36]: I think status here should be accpeted or not.
May be other things you have here can be events from node to mesh

File: systems/node/controller/pkg/server/controller_test.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/ukama/ukama/systems/node/controller/mocks"
+	pb "github.com/ukama/ukama/systems/node/controller/pb/gen"
+	"github.com/ukama/ukama/systems/node/controller/pkg"
+)
+
+const testOrgName = "test-org"
+
+var orgId = uuid.NewV4()
+
+func TestControllerServer_RestartSite(t *testing.T) {
```
Comment (vthakur7f) [inline systems/node/controller/pkg/server/controller_test.go:20]: RestartNode test is missing

File: systems/node/controller/Int.Dockerfile [hunk 8]
```diff
@@ -0,0 +1,3 @@
+FROM 003664043471.dkr.ecr.us-east-1.amazonaws.com/e2e-base-image:latest 
+ENV PROJECT_NAME=controller
+COPY bin/integration /usr/bin/integration
```
Comment (vthakur7f) [inline systems/node/controller/Int.Dockerfile:3]: integration test missing

**Review by vthakur7f — COMMENTED** [review]
Some minor changes required otherwise looks good.

File: systems/node/api-gateway/pkg/rest/router.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+		if err == nil {
+			return
+		}
+	})
+	auth.Use()
+	{
+
+		const cont = "/controllers"
+		controller := auth.Group(cont, "Controllers", "Operations on controllers")
+		controller.POST("/restartSite", formatDoc("restart site in org", "restaring a site within an org "), tonic.Handler(r.postRestartSiteHandler, http.StatusCreated))
```
[the remark this answers is not in the document]
Comment (Brackleycassinga) [reply systems/node/api-gateway/pkg/rest/router.go:109]: Fixed.

[the remark this answers is not in the document]
Comment (Brackleycassinga) [reply systems/node/api-gateway/pkg/rest/router.go:109]: I have addressed the issue following your guideline:
here is the updated endpoints:
```
const cont = "/controllers"
		controller := auth.Group(cont, "Controllers", "Operations on controllers")
		controller.POST("/networks/:network_id/sites/:site_name/restart", formatDoc("Restart a site in an organization", "Restarting a site within an organization"), tonic.Handler(r.postRestartSiteHandler, http.StatusOK))
		controller.POST("/nodes/:node_id/restart", formatDoc("Restart a node", "Restarting a node"), tonic.Handler(r.postRestartNodeHandler, http.StatusOK))
		controller.POST("/networks/:network_id/restart-nodes", formatDoc("Restart multiple nodes within a network", "Restarting multiple nodes within a network"), tonic.Handler(r.postRestartNodesHandler, http.StatusOK))
```

File: systems/node/api-gateway/pkg/rest/router_test.go (hunk 1 above)
Comment (Brackleycassinga) [reply systems/node/api-gateway/pkg/rest/router_test.go:84]: More test added and legacy code removed.

File: systems/node/controller/pb/controller.proto (hunk 3 above)
Comment (Brackleycassinga) [reply systems/node/controller/pb/controller.proto:17]: Addressed

File: systems/node/controller/pb/controller.proto (hunk 4 above)
Comment (Brackleycassinga) [reply systems/node/controller/pb/controller.proto:27]: Addressed

File: systems/node/controller/pb/controller.proto (hunk 5 above)
Comment (Brackleycassinga) [reply systems/node/controller/pb/controller.proto:32]: Addressed

File: systems/node/controller/pkg/server/controller_test.go (hunk 7 above)
Comment (Brackleycassinga) [reply systems/node/controller/pkg/server/controller_test.go:20]: test added

File: systems/common/pb/gen/ukama/node-feeder.pb.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+)
+
+const (
+	// Verify that this generated code is sufficiently up-to-date.
+	_ = protoimpl.EnforceVersion(20 - protoimpl.MinVersion)
+	// Verify that runtime/protoimpl is sufficiently up-to-date.
+	_ = protoimpl.EnforceVersion(protoimpl.MaxVersion - 20)
+)
+
+type NodeUpdateRequest struct {
```
Comment (vthakur7f) [inline systems/common/pb/gen/ukama/node-feeder.pb.go:25]: This is generic message Should have some name like NodeFeederMsg

File: systems/messaging/node-feeder/README.md [hunk 11]
```diff
@@ -1,31 +1,31 @@
...
+Node Feeder sends HTTP request to list of nodes.
 
 ## Using Node Feeder
-To send request to a group of devices send a message to `amq.topic` exchange with routing key `request.cloud.node-feeder`.
+
+To send request to a group of Nodes send a message to `amq.topic` exchange with routing key `request.cloud.node-feeder`.
 The body should have the following format:
 
 ```
 {
```
Comment (vthakur7f) [inline systems/messaging/node-feeder/README.md:11]: this should be proto-buffer, no more json

File: systems/messaging/node-feeder/pkg/listener.go [hunk 12]
```diff
@@ -212,10 +208,9 @@ func (q *QueueListener) isRetryLimitReached(delivery amqp.Delivery) bool {
 	return false
 }
 
-// return error only if it could be fixed by retry
-// malformed request should not be considered as error
+//process nodeUpdateRequest msg also
 func (q *QueueListener) processRequest(delivery amqp.Delivery) error {
-	request := &DevicesUpdateRequest{}
+	request := &NodeUpdateRequest{}
 	err := json.Unmarshal(delivery.Body, request)
```
Comment (vthakur7f) [inline systems/messaging/node-feeder/pkg/listener.go:214]: Again it's protobuffer you receive here of type any.  Look at line 144 https://github.com/ukama/ukama/blob/main/systems/services/msgClient/internal/queue/listener.go
```
e := &pb.Event{
		RoutingKey: d.RoutingKey,
		Msg:        evtAny,
	}
```

evtAny is your NodeFeederMsg:

```
Something like this :
```
	msg := &pb.NodeFeederMessage{
				Target:     c.OrgName + "." + meteData.network + "." + meteData.site + "." + n,
				HTTPMethod: "POST",
				Path:       "/v1/configd/config",
				Msg:        anyMsg,
			}
```
Again you can have a loog at configurator configstore package to see how the protbuffer are packed and you have to follow reverse process here to unpack.

**Review by vthakur7f — CHANGES_REQUESTED** [review]

File: systems/messaging/node-feeder/README.md (hunk 11 above)
Comment (Brackleycassinga) [reply systems/messaging/node-feeder/README.md:21]: addressed

File: systems/messaging/node-feeder/pkg/listener.go (hunk 12 above)
Comment (Brackleycassinga) [reply systems/messaging/node-feeder/pkg/listener.go]: Fixed

**Comment by Brackleycassinga** [comment]
Close because all code are already in node-feede-updates

**Pull request closed without merging**